Bindable layouts

andreinitescu opened the pull request:
### Description of Change ###

It allows any `Layout<View>` to generate its content (child views) by binding to an items source, with option to set item template or item template selector:

```
<StackLayout BindableLayout.ItemsSource="{Binding Items}" />

<StackLayout BindableLayout.ItemsSource="{Binding Items}"
             BindableLayout.ItemTemplate="{Binding MyItemTemplate}" />

<StackLayout BindableLayout.ItemsSource="{Binding Items}"
             BindableLayout.ItemTemplateSelector="{Binding MyItemTemplateSelector}" />
```

In the first example, where neither item template nor item template selector are set, every item is a `Label`. If both item template and selector are set, the item template is used. If `Items` property is set, it is ignored.

### Issues Resolved ### 
<!-- Please use the format "fixes #xxxx" for each issue this PR addresses -->

- fixes #1718

### API Changes ###

Added static class BindableLayout with following attached properties to `Layout<View>`:

 - public IEnumerable ItemsSource{ get; set; }
 - public DataTemplate ItemTemplate { get; set; }
 - public DataTemplateSelector ItemTemplateSelector { get; set; }
 - private BindableLayoutController BindableLayoutController { get; set; }

Added private class BindableLayoutController which encapsulates updating layout children based on changes on the collection, item template or item template selector.

### Platforms Affected ### 
<!-- Please list all platforms affected by these changes -->

- Core/XAML (all platforms)

### Testing Procedure ###

I added a gallery page (see screenshot below) which helps with testing all operations on a collection.

###  Before/After Screenshots ###

This change doesn't affect layout appearance, no matter the layout.  
However I attached a screenshot which shows the gallery page:
![image](https://user-images.githubusercontent.com/743918/46918637-e1a14f00-cfdc-11e8-9fa1-bc69a05df172.png)

The buttons should be self explanatory. "Replace" turns integers into chars to demonstrate replacing items in the observable collection. "Move" moves first item to last.

### PR Checklist ###

- [x] Has automated tests (See `BindableLayoutTests` unit tests) <!-- (if tests are omitted or manual, state reason in description) -->
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

StephaneDelcroix commented on Xamarin.Forms.Core/BindableLayout.cs:
```diff
@@ -0,0 +1,261 @@
+	public static class BindableLayout
+	{
+		public static readonly BindableProperty ItemsSourceProperty =
+			BindableProperty.Create("ItemsSource", typeof(IList), typeof(Layout<View>), default(IList),
```
`CreateAttached`

StephaneDelcroix replied:
should be `IEnumerable`

andreinitescu replied:
Ooops! Done!

StephaneDelcroix commented on Xamarin.Forms.Core/BindableLayout.cs:
```diff
@@ -0,0 +1,261 @@
+			BindableProperty.Create("ItemTemplateSelector", typeof(DataTemplateSelector), typeof(Layout<View>),
+				propertyChanged: (b, o, n) => { GetController(b).ItemTemplateSelector = (DataTemplateSelector)n; });
+
+		public static readonly BindableProperty ControllerProperty =
```
shouldn't be `public`

andreinitescu replied:
If it's not public it completely eliminates the possibility to override item view setup. 
I tried to describe in detail in the comments in the code example above the reasoning behind controller. Maybe I wasn't clear enough? 
I'd like to hear more thoughts. Thanks!

andreinitescu replied:
Should I maybe provide an example to show how overriding the controller makes sense?

andreinitescu replied:
The following two comments you made are related to this too.

andreinitescu replied:
Or maybe I make it private for now and open a separate enhancement PR to make it public? 
Would this be quicker in order to get this functionality merged and released?

andreinitescu replied:
Done, I removed public. I think it's a missed opportunity

StephaneDelcroix commented on Xamarin.Forms.Core/BindableLayout.cs:
```diff
@@ -0,0 +1,261 @@
+			return (DataTemplateSelector)b.GetValue(ItemTemplateSelectorProperty);
+		}
+
+		public static BindableLayoutController GetController(BindableObject b)
```
shouldn't be `public`

andreinitescu replied:
Done, I removed public. I think it's a missed opportunity

StephaneDelcroix commented on Xamarin.Forms.Core/BindableLayout.cs:
```diff
@@ -0,0 +1,261 @@
+			return (BindableLayoutController)b.GetValue(ControllerProperty);
+		}
+
+		public static void SetController(BindableObject b, BindableLayoutController value)
```
shouldn't be `public`, naming is inconsistent

andreinitescu replied:
Which naming are you referring to?

andreinitescu replied:
Done, I removed public. I think it's a missed opportunity

StephaneDelcroix commented on Xamarin.Forms.Core/BindableLayout.cs:
```diff
@@ -0,0 +1,261 @@
+		}
+	}
+
+	public class BindableLayoutController
```
no need to be `public`

andreinitescu replied:
Done, I removed public. I think it's a missed opportunity

andreinitescu commented:
@StephaneDelcroix All requested changes done

rmarinho commented on Xamarin.Forms.Core/BindableLayout.cs:
```diff
@@ -0,0 +1,262 @@
+			BindableProperty.CreateAttached("ItemTemplateSelector", typeof(DataTemplateSelector), typeof(Layout<View>), default(DataTemplateSelector),
+				propertyChanged: (b, o, n) => { GetController(b).ItemTemplateSelector = (DataTemplateSelector)n; });
+
+		static readonly BindableProperty ControllerProperty =
```
Should this be `BindableLayoutControllerProperty` ?

andreinitescu replied:
Done!

sl154672 commented:
Red fish Blue fish = green fistch

Sent from my iPhone

> On Oct 12, 2561 BE, at 1:06 PM, Rui Marinho <notifications@github.com> wrote:
> 
> @rmarinho requested changes on this pull request.
> 
> Hi @andreinitescu thanks for your contribution. We will need some unit tests.
> UI Tests and a page on the gallery will also be great so we can try this out and test it.
> 
> In Xamarin.Forms.Core/BindableLayout.cs:
> 
> > +		View CreateItemView(object item, int index, DataTemplate dataTemplate)
> +		{
> +			if (dataTemplate != null)
> +			{
> +				var view = (View)dataTemplate.CreateContent();
> +				view.BindingContext = item;
> +				OnItemViewCreated(view, item, index);
> +				return view;
> +			}
> +			else
> +			{
> +				return new Label() { Text = item?.ToString() };
> +			}
> +		}
> +
> +		void ItemsSource_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
> Can we drop the _
> 
> In Xamarin.Forms.Core/BindableLayout.cs:
> 
> > +{
> +	public static class BindableLayout
> +	{
> +		public static readonly BindableProperty ItemsSourceProperty =
> +			BindableProperty.CreateAttached("ItemsSource", typeof(IEnumerable), typeof(Layout<View>), default(IList),
> +				propertyChanged: (b, o, n) => { GetController(b).ItemsSource = (IEnumerable)n; });
> +
> +		public static readonly BindableProperty ItemTemplateProperty =
> +			BindableProperty.CreateAttached("ItemTemplate", typeof(DataTemplate), typeof(Layout<View>), default(DataTemplate),
> +				propertyChanged: (b, o, n) => { GetController(b).ItemTemplate = (DataTemplate)n; });
> +
> +		public static readonly BindableProperty ItemTemplateSelectorProperty =
> +			BindableProperty.CreateAttached("ItemTemplateSelector", typeof(DataTemplateSelector), typeof(Layout<View>), default(DataTemplateSelector),
> +				propertyChanged: (b, o, n) => { GetController(b).ItemTemplateSelector = (DataTemplateSelector)n; });
> +
> +		static readonly BindableProperty ControllerProperty =
> Should this be BindableLayoutControllerProperty ?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

StephaneDelcroix commented:
@rmarinho UITests aren't required, it can all be unit-tested

andreinitescu commented:
@rmarinho @StephaneDelcroix 

Updates:
- Changed controller property name as you suggested 
- Added gallery page
- Added unit tests
- Updated the ticket description and added screenshot

@rmarinho Regarding your suggestion to updated description with relevant scenario for controller:
Since now setting controller is private, advanced scenario is not applicable anymore.
Once this ticket gets merged, I will open a separate ticket to discuss making controller public and describing relevant scenarios.

andreinitescu commented on Xamarin.Forms.Core/BindableLayout.cs:
```diff
@@ -0,0 +1,262 @@
+			}
+		}
+
+		void ItemsSource_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
```
Done!

andreinitescu commented:
@StephaneDelcroix @rmarinho @hartez Would appreciate your feedback, thanks!

andreinitescu commented:
@StephaneDelcroix @rmarinho @hartez  Can someone say why this failed?

_"Xamarin Forms OSX — #pull-4052 - (2121984) failed"_

I don't have access to see it.
I checked it on my Mac and the control gallery app runs fine and bindable layout works perfectly.

andreinitescu commented:
build

andreinitescu commented:
@StephaneDelcroix @rmarinho @hartez Would appreciate your feedback, thanks!

rmarinho commented:
Just tested, works fine for me. All my feedback was address. Good job.

andreinitescu commented:
@StephaneDelcroix Please let me know if there's anything you need from me. Thanks.

andreinitescu commented:
@StephaneDelcroix  Please let me know if there's anything you need from me. Thanks.

andreinitescu commented:
@StephaneDelcroix  Please let me know if there's anything you need from me. Thanks.
